vault/barrier_aes_gcm.go

```diff
@@ -1049,3 +1106,40 @@ func (b *AESGCMBarrier) Keyring() (*Keyring, error) {
 
 	return b.keyring.Clone(), nil
 }
+
+func (b *AESGCMBarrier) encryptTracked(ctx context.Context, keyring *Keyring, path string, term uint32, gcm cipher.AEAD, buf []byte) ([]byte, error) {
+	ct, err := b.encrypt(path, term, gcm, buf)
+	if err != nil {
+		return nil, err
+	}
+	// Increment the local encryption count, and track metrics
+	ops := atomic.AddInt64(&keyring.LocalEncryptions, 1)
+	metrics.IncrCounterWithLabels(barrierEncryptsMetric, 1, termLabel(term))
+
+	// time.Now is expensive, don't check time based rotation on every operation
+	if b.shouldRotate(ops, keyring) {
+		b.l.Lock()
+		defer b.l.Unlock()
+		// Retest the precondition
```

Comment:
Why does the precondition need to be retested?

Reply:
There is a potential race here where two encryptions happen close enough together that they both want to trigger rotation.  If you don't retest you can rotate twice.

Reply:
This got cleaned up as I brought over the changes made in ENT to here.